zio-http/shared/src/main/scala/zio/http/codec/HttpContentCodec.scala

```diff
@@ -0,0 +1,195 @@
+    lookup(contentType) match {
+      case Some(codec) =>
+        request.body.asChunk.flatMap { bytes =>
+          ZIO.fromEither(codec.decode(bytes))
+        }
+      case None        =>
+        ZIO.fail(throw new IllegalArgumentException(s"No codec found for content type $contentType"))
+    }
+  }
+  def decodeResponse(response: Response): Task[A] = {
+    val contentType =
+      response.header(Header.ContentType).getOrElse(Header.ContentType(MediaType.application.`json`))
+    lookup(contentType) match {
+      case Some(codec) =>
+        response.body.asChunk.flatMap { bytes =>
+          ZIO.fromEither(codec.decode(bytes))
+        }
+      case None        =>
+        ZIO.fail(throw new IllegalArgumentException(s"No codec found for content type $contentType"))
+    }
+  }
+
+  def encode(value: A): Either[String, Body] =
+    choices.headOption match {
+      case Some((mediaType, codec)) =>
+        Right(Body.fromChunk(codec.encode(value)).contentType(mediaType))
+      case None                     =>
+        Left("No codec found")
+    }
+
+  def only(mediaType: MediaType): HttpContentCodec[A] =
+    HttpContentCodec(
+      ListMap(
+        mediaType -> lookup(mediaType)
+          .getOrElse(
+            throw new IllegalArgumentException(s"MediaType $mediaType is not supported by $self"),
+          ),
+      ),
+      schema,
+    )
+
+  def only(mediaType: Option[MediaType]): HttpContentCodec[A] = {
+    mediaType match {
+      case Some(mediaType) =>
+        HttpContentCodec(
+          ListMap(
+            mediaType -> lookup(mediaType)
+              .getOrElse(
+                throw new IllegalArgumentException(s"MediaType $mediaType is not supported by $self"),
+              ),
+          ),
+          schema,
+        )
+      case None            =>
+        self
+    }
+  }
+
+  def onlyOrNone(mediaType: MediaType): Option[HttpContentCodec[A]] =
+    if (choices.contains(mediaType)) Some(only(mediaType))
+    else None
+
+  def lookup(mediaType: MediaType): Option[BinaryCodec[A]] =
+    choices.collectFirst { case (mt, codec) if mt.matches(mediaType) => codec }
+
+  def lookup(contentType: Header.ContentType): Option[BinaryCodec[A]] =
+    lookup(contentType.mediaType)
+
+  private[http] lazy val defaultMediaType: MediaType =
+    choices.headOption.map(_._1).getOrElse {
+      throw new IllegalArgumentException(s"No codec defined")
+    }
+
+  lazy val defaultCodec: BinaryCodec[A] = choices.headOption.map(_._2).getOrElse {
+    throw new IllegalArgumentException(s"No codec defined")
+  }
+}
+
+object HttpContentCodec {
+  implicit def fromSchema[A](implicit schema: Schema[A]): HttpContentCodec[A] = {
+    json[A] ++ protobuf[A] ++ text[A]
+  }
+
+  object json {
+    def apply[A](implicit schema: Schema[A]): HttpContentCodec[A]               = fromSchema[A]
+    implicit def fromSchema[A](implicit schema: Schema[A]): HttpContentCodec[A] = {
```

Comment:
😱  I don't like multiple implicit conversions. Well, I don't like implicit conversions, and I REALLY don't like multiple implicit conversions. 😄 

Let's remove all of this and just say if you want that much control, you are on your own.

Then maybe we can turn `HttpContentCodec.json.apply` into `HttpContentCodec.jsonOnly`, etc.